### PR TITLE
Avoid unnecessary is same

### PR DESCRIPTION
**Pull request by pereckerdal**
@punchfox pointed out in #29 that this doesn't work.

### CODE REVIEW

**Comment by pereckerdal** [comment]
@punchfox @fxb @emil-e

File: test/src/test_encoded_value.cpp [hunk 1]
```diff
@@ -105,6 +105,46 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(
...
+BOOST_AUTO_TEST_CASE_TEMPLATE(
+    json_encoded_value_should_compare_for_equality,
+    encoded_value_type,
+    encoded_value_types) {
+  encoded_value_type a("null");
+  encoded_value_type b("1");
+  BOOST_CHECK(a == a);
+  BOOST_CHECK(!(a == b));
+  BOOST_CHECK(!(a != a));
+  BOOST_CHECK(a != b);
```
Comment (fxb) [inline test/src/test_encoded_value.cpp]: Could use BOOST_CHECK_EQUAL and BOOST_CHECK_NE here. But possibly the way you did it, it is easier to understand. Also the third case wouldn't be possible to check that way.

File: include/spotify/json/codec/array.hpp [hunk 2]
```diff
@@ -125,10 +125,15 @@ class array_t final {
...
+      std::is_convertible<
           typename T::value_type,
           typename std::decay<codec_type>::type::object_type>::value,
-      "Array container type must match inner codec type");
+          "Array container type must be convertible to inner codec type");
+  static_assert(
+      std::is_convertible<
+          typename std::decay<codec_type>::type::object_type,
+          typename T::value_type>::value,
+          "Inner codec type must be convertible to array container type");
```
Comment (fxb) [inline include/spotify/json/codec/array.hpp]: Indentation should be one level less for the static_assert message.

File: include/spotify/json/encoded_value.hpp [hunk 3]
```diff
@@ -209,5 +209,13 @@ inline std::ostream &operator <<(std::ostream &stream, const encoded_value &valu
   return stream;
 }
 
+inline bool operator==(const encoded_value_ref &a, const encoded_value_ref &b) {
+  return a.size() == b.size() && std::memcmp(a.data(), b.data(), a.size()) == 0;
```
Comment (fxb) [inline include/spotify/json/encoded_value.hpp]: Also check for a.data() == b.data() and short-circuit it? Maybe that's overkill. Maybe std::memcmp already does it (in some implementations?) !?

File: test/src/test_encoded_value.cpp (hunk 1 above)
Comment (pereckerdal) [inline test/src/test_encoded_value.cpp]: I like to make it obvious that it's using the operators I want it to use here. It's not obvious that `BOOST_CHECK_NE` actually uses `operator!=`

File: include/spotify/json/codec/array.hpp (hunk 2 above)
Comment (pereckerdal) [inline include/spotify/json/codec/array.hpp]: will fix

**Comment by pereckerdal** [comment]
@fxb I fixed most of your comments

**Comment by fxb** [comment]
@pereckerdal: Feel free to merge when builds are green!

**Comment by punchfox** [comment]
👍 Nice

**Pull request merged**